sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.parquet
+
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, Expression, NamedExpression}
+import org.apache.spark.sql.catalyst.planning.{PhysicalOperation, ProjectionOverSchema, SelectedField}
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, LogicalRelation}
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * Prunes unnecessary Parquet columns given a [[PhysicalOperation]] over a
+ * [[ParquetRelation]]. By "Parquet column", we mean a column as defined in the
+ * Parquet format. In Spark SQL, a root-level Parquet column corresponds to a
+ * SQL column, and a nested Parquet column corresponds to a [[StructField]].
+ */
+private[sql] object ParquetSchemaPruning extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan =
+    plan transformDown {
+      case op @ PhysicalOperation(projects, filters,
+          l @ LogicalRelation(hadoopFsRelation @ HadoopFsRelation(_, partitionSchema,
+            dataSchema, _, parquetFormat: ParquetFileFormat, _), _, _)) =>
+        val projectionFields = projects.flatMap(getFields)
+        val filterFields = filters.flatMap(getFields)
+        val requestedFields = (projectionFields ++ filterFields).distinct
+
+        // If [[requestedFields]] includes a proper field, continue. Otherwise,
+        // return [[op]]
+        if (requestedFields.exists { case (_, optAtt) => optAtt.isEmpty }) {
+          val prunedSchema = requestedFields
+            .map { case (field, _) => StructType(Array(field)) }
+            .reduceLeft(_ merge _)
```

Comment:
Can the merging order of the fields affect the comparison later in line 57?

For example, depending on the merging order, if we require `a.field1` and `a.field2`, is it possibly we may have `StructType(StructField("a", StructType(StructField("field1") :: StructField("field2") :: Nil)) :: Nil)` or  `StructType(StructField("a", StructType(StructField("field2") :: StructField("field1") :: Nil)) :: Nil)` as merged schema?

Once we have the later, and the data schema is the first, at line 57 we may think they are different?

Reply:
@viirya @mallman 

If I may add here, given a schema:
```
root
| - a: StructType
|    | - f1: Int
|    | - f2: Int
```
and a selection `df.select("a.f1", "a.f2")` vs `df.select("a.f2", "a.f1")`
will produce different requiredSchema fed into buildReader upon some action.

In the first case it will be `StructType( StructField( "a", StructType(StructField("f1") :: StructField("f2") :: Nil)) :: Nil)` and in the second `StructType( StructField( "a", StructType(StructField("f2") :: StructField("f1") :: Nil)) :: Nil)` 

which means that the original schema is different from the one that is required for the second case. But as long as fields f1 and f2 are splitted (can be read without reading the other), it's remappable on the data source level.

VK

Reply:
We should add related tests for such cases.

Reply:
> Can the merging order of the fields affect the comparison later in line 57?

Not the merging order, no. However the selection order can, leading to unnecessary "pruning". I will address this issue and add a test case to verify it.

Reply:
I've updated the logic for comparing the original schema to the pruned schema:

https://github.com/apache/spark/blob/52fddc181f32726cea1dd12a23ebf7201986be01/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala#L53-L57

The following test validates that the selection order is irrelevant in comparing the pruned schema to the original schema:

https://github.com/apache/spark/blob/52fddc181f32726cea1dd12a23ebf7201986be01/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala#L82-L107